estimators/all_estimators.py

```diff
@@ -0,0 +1,279 @@
+
+
+def _calculate_propensity_score(param_method, data, y_treated_hat, y_control_hat, t, y):
+    """ Calculating Propensity Score and return the treatment effect.
+
+    :param param_method: dictionary.
+    :param data: tf.data.Dataset.
+    :param y_treated_hat: [], predicted outcome if treated.
+    :param y_control_hat: [], predicted outcome if control.
+    :param t: [], treatment assignment.
+    :param y: [], observed outcome.
+    :return: tau (int, treat effect), var (int, variance)
+    """
+    t = t.ravel()
+    y = y.ravel()
+
+    # Calculating the Propensity Score.
+    if param_method['learn_prop_score']:
+        data_all = data.dataset_all_ps
+        # Propensity score using a logistic regression.
+        prop_score = param_method['prop_score']
+        prop_score.fit(data_all)
+        e = prop_score.predict_proba(data_all)
+        # Removing extreme values
+        # print('y shaoes', y.shape, type(y))
+        g = e[:, 1].ravel()
+        y = _truncate_by_g(y.reshape(-1, 1), g, 0.005)
+        t = _truncate_by_g(t, g, 0.005)
+        y_treated_hat = _truncate_by_g(y_treated_hat, g, 0.005)
+        y_control_hat = _truncate_by_g(y_control_hat, g, 0.005)
+        e = _truncate_by_g(e, g, 0.005)
+
+        # Estimating the treatment effect.
+        y_dif = (y_treated_hat - y_control_hat)
+        sample_size = len(y)
+        residual_treated = (t * (y - y_treated_hat))
+        residual_control = ((1 - t) * (y - y_control_hat))
+
+        residual_treated = (residual_treated / e[:, 1].ravel())
+        residual_control = (residual_control / e[:, 0].ravel())
+
+    else:
+        # Estimating the treatment effect.
+        y_dif = (y_treated_hat - y_control_hat)
+        sample_size = len(y)
+        residual_treated = (t * (y - y_treated_hat))
+        residual_control = ((1 - t) * (y - y_control_hat))
+        residual_treated = (residual_treated / 0.5)
+        residual_control = (residual_control / 0.5)
+
+    # Calculating the treatment effect.
+    residual_dif = (residual_treated - residual_control)
+    tau_estimated = np.mean(np.add(y_dif, residual_dif))
+
+    # Variance.
+    var = np.add(y_dif, residual_dif)
+    var = np.subtract(var, tau_estimated)
+    var = var ** 2
+    var = var.sum() / (sample_size - 1)
+    var = var / sample_size
+
+    return tau_estimated, var
+
+
+def _truncate_by_g(attribute, g, level=0.005):
+    """ Remove samples with extreme propensity score values for stability.
+
+    Remove rows with too low or too high g values. attribute and g must have same dimensions.
+    :param attribute: [], array to be filtered.
+    :param g: [], array used to filter.
+    :param level: float, limites.
+    :return: filted attribute column
+    """
+    assert len(attribute) == len(g), 'Dimensions must be the same!' + str(len(attribute)) + ' and ' + str(len(g))
+    keep_these = np.logical_and(g >= level, g <= 1. - level)
+    return attribute[keep_these]
```

Comment:
Are you sure you want to remove the small values? An alternative would be to set `attribute = np.maximum(attribute, level)

Reply:
Ok! I changed so now we have a minimum / maximum values for the propensity score [level, 1-level]. The items associated with small values are no longer removed.